### PR TITLE
Switch to cmake for builds

### PR DESCRIPTION
**Pull request by berkus**
* Cleanup code in general
* Remove custom types
* Remove autoupdater
* Add conan-based build

Obsoletes #22 - this one is ready to merge once tested successfully.

### CODE REVIEW

File: Telegram/SourceFiles/config.h [hunk 1]
```diff
@@ -247,7 +247,7 @@ w/CVnbwQOw0g5GBwwFV3r0uTTvy44xx8XXxk+Qknu4eBCsmrAFNnAgMBAAE=\n\
 #ifdef CUSTOM_API_ID
 #include "../../../TelegramPrivate/custom_api_id.h" // Custom API id and API hash
 #else
-static const int32 ApiId = 17349;
+static const int32_t ApiId = 17349;
```
[the remark this answers is not in the document]
Comment (ForNeVeR) [reply Telegram/SourceFiles/config.h]: @leha-bot, please file an issue for that!

File: README.md [hunk 2]
```diff
@@ -28,7 +39,6 @@ The source code is published under GPLv3 with OpenSSL exception, the license is
 * liblzma ([public domain](http://tukaani.org/xz/))
 * Google Breakpad ([License](https://chromium.googlesource.com/breakpad/breakpad/+/master/LICENSE))
 * Google Crashpad ([Apache License 2.0](https://chromium.googlesource.com/crashpad/crashpad/+/master/LICENSE))
-* GYP ([BSD License](https://github.com/bnoordhuis/gyp/blob/master/LICENSE))
 * Ninja ([Apache License 2.0](https://github.com/ninja-build/ninja/blob/master/COPYING))
```
[the remark this answers is not in the document]
Comment (berkus) [reply README.md:51]: Gyp? no

✅

File: Telegram/SourceFiles/mainwindow.h [hunk 3]
```diff
@@ -383,11 +355,11 @@ class NetworkSettingsWindow : public PreLaunchWindow {
 
 public:
 
-	NetworkSettingsWindow(QWidget *parent, QString host, quint32 port, QString username, QString password);
+	NetworkSettingsWindow(QWidget *parent, QString host, uint32_t port, QString username, QString password);
 
 signals:
 
-	void saved(QString host, quint32 port, QString username, QString password);
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/mainwindow.h]: This is fine :finnadie:

File: Telegram/SourceFiles/media/media_audio.h [hunk 4]
```diff
@@ -121,17 +121,17 @@ class Mixer : public QObject, private base::Subscriber {
 public:
 	Mixer();
 
-	void play(const AudioMsgId &audio, int64 position = 0);
-	void play(const AudioMsgId &audio, std::unique_ptr<VideoSoundData> videoData, int64 position = 0);
+	void play(const AudioMsgId &audio, int64_t position = 0);
+	void play(const AudioMsgId &audio, std::unique_ptr<VideoSoundData> videoData, int64_t position = 0);
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/media/media_audio.h]: What if it's moved?

File: Telegram/SourceFiles/mtproto/core_types.h [hunk 5]
```diff
@@ -22,23 +22,24 @@ Copyright (c) 2014-2017 John Preston, https://desktop.telegram.org
 
 #include "core/basic_types.h"
 #include "base/flags.h"
+#include <cstdint>
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/mtproto/core_types.h:25]: We want to get rid of basic_types.h

✅

File: Telegram/Telegram.plist [hunk 6]
```diff
@@ -3,24 +3,28 @@
 <plist version="1.0">
 <dict>
 	<key>CFBundleExecutable</key>
-	<string>$(PRODUCT_NAME)</string>
+	<string>${MACOSX_BUNDLE_EXECUTABLE_NAME}</string>
 	<key>CFBundleGetInfoString</key>
-	<string>Telegram Desktop messaging app</string>
+	<string>pro.cxx Telegram messaging app</string>
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/Telegram.plist:8]: Ok

File: Telegram/SourceFiles/core/version.h [hunk 7]
```diff
@@ -20,11 +20,12 @@ Copyright (c) 2014-2017 John Preston, https://desktop.telegram.org
...
 
-#include "core/utils.h"
+#include <cstdint>
+// #include "core/utils.h"
 
 #define BETA_VERSION_MACRO (0ULL)
 
 constexpr int AppVersion = 1001023;
-constexpr str_const AppVersionStr = "1.1.23";
+constexpr str_const AppVersionStr = "1.1.23-procxx";
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/core/version.h:29]: Created #34 as follow-up.

✅

File: Telegram/Telegram.plist (hunk 6 above)
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/Telegram.plist:8]: Created #34 as follow-up.

✅

File: Telegram/SourceFiles/apiwrap.cpp [hunk 8]
```diff
@@ -1405,13 +1405,13 @@ void ApiWrap::gotWebPages(ChannelData *channel, const MTPmessages_Messages &msgs
 	}
 
 	if (!v) return;
-	QMap<uint64, int32> msgsIds; // copied from feedMsgs
-	for (int32 i = 0, l = v->size(); i < l; ++i) {
+	QMap<uint64_t, int32_t> msgsIds; // copied from feedMsgs
+	for (int32_t i = 0, l = v->size(); i < l; ++i) {
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/apiwrap.cpp]: Created #35 as follow-up.

✅

File: Telegram/SourceFiles/boxes/abstract_box.cpp [hunk 9]
```diff
@@ -104,7 +104,7 @@ void BoxContent::onDraggingScrollDelta(int delta) {
 }
 
 void BoxContent::onDraggingScrollTimer() {
-	auto delta = (_draggingScrollDelta > 0) ? qMin(_draggingScrollDelta * 3 / 20 + 1, int32(MaxScrollSpeed)) : qMax(_draggingScrollDelta * 3 / 20 - 1, -int32(MaxScrollSpeed));
+	auto delta = (_draggingScrollDelta > 0) ? qMin(_draggingScrollDelta * 3 / 20 + 1, int32_t(MaxScrollSpeed)) : qMax(_draggingScrollDelta * 3 / 20 - 1, -int32_t(MaxScrollSpeed));
```
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/boxes/abstract_box.cpp]: Created #36 as a follow-up.

✅

File: Telegram/SourceFiles/config.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (berkus) [reply Telegram/SourceFiles/config.h]: Filed #37 as follow-up item.

✅

**Review by crackedmind — APPROVED** [review]

**Pull request merged**